images/virtualization-artifact/pkg/controller/vmip/internal/iplease_handler.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2024 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package internal
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/deckhouse/virtualization-controller/pkg/controller/common"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/service"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/vmip/internal/state"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/vmip/internal/util"
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+)
+
+const IpLeaseHandlerName = "IPLeaseHandler"
+
+type IPLeaseHandler struct {
+	client    client.Client
+	logger    logr.Logger
+	ipService *service.IpAddressService
+}
+
+func NewIPLeaseHandler(client client.Client, logger logr.Logger, ipAddressService *service.IpAddressService) *IPLeaseHandler {
+	return &IPLeaseHandler{
+		client:    client,
+		logger:    logger.WithValues("handler", IpLeaseHandlerName),
+		ipService: ipAddressService,
+	}
+}
+
+func (h IPLeaseHandler) Handle(ctx context.Context, state state.VMIPState) (reconcile.Result, error) {
+	vmip := state.VirtualMachineIP()
+	vmipStatus := state.VirtualMachineIPStatus()
+
+	lease, err := state.VirtualMachineIPLease(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	switch {
+	case lease == nil && vmipStatus.Address != "":
+		h.logger.Info("Lease by name not found: waiting for the lease to be available")
+		return reconcile.Result{}, nil
+
+	case lease == nil:
+		h.logger.Info("No Lease for VirtualMachineIP: create the new one", "type", vmip.Spec.Type, "address", vmip.Spec.StaticIP)
+		return h.createNewLease(ctx, state)
+
+	case lease.Status.Phase == "":
+		h.logger.Info("Lease is not ready: waiting for the lease")
+		return reconcile.Result{Requeue: true, RequeueAfter: 2 * time.Second}, nil
```

Comment:
We replace this requeue with proper UpdateFunc in VirtualMachineIPAddressLease watch setup. May be add TODO for this transition in the future.

Reply:
Done